Components: Export reusable components to a separate file + global

ryelle opened the pull request:
This PR builds on #287 so that we can use the report filters in third-party development. This is a chain of PRs working towards extensibility in WC Admin, #254 . In particular, this exports our components on the `wc` global, so they can be used by anyone.

I've added a new file, `components/index.js`, which serves as a new entry point in webpack. This builds to the `dist/components.js` file, which when loaded populates `wc.components` with the exported components. This new file is registered & added as a dependency of the wc-admin app, `wc-components`.

Next, I updated webpack.config.js, so we can use `@woocommerce/components` like `@wordpress/components`, and updated all component imports in non-component folders to pull from  `@woocommerce/components`. This ensures we're always pulling components from the global  `wp.components` rather than importing component code back into the index/embedded files.

Inside `component/*` files, we import from `component/*` paths, to prevent circular dependencies.

I started to break apart "complex" and "simple" components, thinking that maybe we'd export these in different files, but I don't think we need to worry about that yet - and if we do, we should talk about what that distinction means.

**To test**

- Build the project, `npm run build` or `npm start`
- See that new files are added to `dist/`: `dist/components.js` + `dist/css/components.css`
- Run the tests to make sure they still work
- Check on the app in browser to make sure there are no UI changes
- In the browser, you can also see that `wc.components` exists - and you can access components, for example `wc.components.Card`.

ryelle committed:
Importing all components from a root components file

ryelle committed:
Update Ratings component to avoid circular dependencies, fix tests

ryelle committed:
Export components on `wc.components`, use this for importing

ryelle committed:
Move react-dates initialize to the components file

ryelle commented:
> I pulled this down and pages don't respond to new query params, eg ?filter=compare. I have to go now, so I didn't look into it, but will come back if I have time tonight

I've rebased the branch on master, and it seems to work -- though I'm now seeing an issue where updating filters is using the path `/analytics?...`  rather than the correct report path. I'll look into this today too.

ryelle committed:
Push query changes to history

Fixes an issues where native a links do not update the query in `history`

ryelle committed:
Update test config for new @woocommerce/components path

timmyc commented on client/components/card/index.js:
```diff
@@ -10,7 +10,7 @@ import PropTypes from 'prop-types';
  */
 import './style.scss';
-import { EllipsisMenu } from '../ellipsis-menu';
+import { EllipsisMenu } from 'components/ellipsis-menu';
```
So this is an example of not importing from `@woocommerce/components` to prevent circular dependencies?

ryelle replied:
Yeah - because this file is imported into `@woocommerce/components`, we don't want to import from `@woocommerce/components`. Far as I can tell it doesn't break or anything, but I think it might slow down build time.

justinshreve replied:
Is there a place we could document that we should import from `components/` directly from within the components directory? I feel like it would be an easy mistake to introduce later, even if it doesn't break anything.

timmyc commented on client/components/index.js:
```diff
@@ -0,0 +1,33 @@
+export { default as SegmentedSelection } from './segmented-selection';
+export { default as SplitButton } from './split-button';
+
+// Complex components
```
Complex components are ones that require more data-driven props like a product object? Is this akin to blocks vs components in Calypso?

ryelle replied:
The complex components were more likely to be made up of other components, or export multiple components (like `TableCard` or `EllipsisMenu`), but ultimately I couldn't make a great distinction between the two, so I just left it like this. We can merge all this back into a single alphabetized list, or try to formalize a difference. This might get more complicated as we get components that also use api requests.

justinshreve replied:
I'd vote for just keeping it simple and alphabetizing a single list. It'll be easy to add a new one without having to think about which group it goes under.

timmyc reviewed (commented):
Really like the globals approach vs Learna/npm for 3pd - seems to me there is less of a barrier to entry and more akin to interacting with other WP core js libraries ( customizer/media modal etc ). Do you think there are any negative drawbacks to going this route vs the lerna approach? Seems the only one might be the ability to use the Woo components outside of the context of wp-admin, but I suppose if that need arises we could add lerna in later.

Couple quick questions in the code, but looking great and testing well for me.

justinshreve commented:
There still seems to be an issue with the filter in this branch. Go to the `Analytics > Products` page first, and update a filter. Then go to the `Analytics > Orders` page and update a filter. It updates on the products path. This doesn't seem to be happening on master.

ryelle commented:
> Do you think there are any negative drawbacks to going this route vs the lerna approach?

Like you said, if someone wants to use it outside of wp-admin it would be tricky (they could use the built `components.js` file directly, but that's definitely not a "good pattern" in the world of JS apps). Build tool setup is a bit trickier because we need to set up aliases possibly in multiple places, but that's a one-time issue when setting up webpack & Jest, and we can document that well - maybe release a "starter plugin" with that all set up.

With the `wc.components` global, we can also be sure that devs are not loading multiple copies of our components (which they could via Lerna, and it could end up being different versions of components 😱 ). This does mean that devs can't control which version of components they get - if we update with a plugin release, everyone gets the new components - so we'll want components to be pretty stable by release time.

timmyc commented:
I can reproduce the issue @justinshreve mentioned above too, and confirmed it doesn't happen on master.

ryelle committed:
Update chart components import

ryelle commented:
@justinshreve @timmyc I think this has something to do with the history object not being totally up-to-date, so that the query/path we get from it is "stuck" on the previous page. you can also replicate a bug by selecting a filter - say, on Orders, "show top orders by items sold", then going to a different page, Products, the filter will still be set to "show top products by items sold".

Interestingly, I can "fix" this just by adding a listener to the history object, changing `lib/history.js`:

```js
import { createHashHistory } from 'history';
const history = createHashHistory();
history.listen( () => {} );
export default history;
```

But that's a hack of a solution 😕 

Of course, we could probably fix this by bringing back the query/path params & passing them through `ReportFilters`... I'm not a huge fan of passing the props down only to pass them on to children components (DatePicker etc), but I suppose that would also make ReportFilters more testable. 

What do you think, are there any other options that I'm missing?

emdashcodes reviewed (approved):
Besides `getQuery()` or passing via props, the only other solution I can think of is looking at `window.location.hash` directly and pulling out the query string part of that.. but that feels hacky as well. Passing via prop might be the best.

Either way, I don't think we need to block the rest of this PR. Approach looks good, and is testing well for me. Maybe just add a TODO or file an issue for the history issue and we can move forward with this?

ryelle committed:
Merge simple/complex & alphabetize by first exported name

ryelle committed:
Add a readme with info about how to call these components

ryelle commented:
I've updated the component export to merge the two lists, and added a README to document the importing behavior.

I've also added #305 to track the history/linking issues, so I can merge this PR now & I think I'll just add the path/query back into passed props.

timmyc commented:
+1 for merging this in
